### PR TITLE
feat: add a simple rate limiter for graphql

### PR DESCRIPTION
**Pull request by jeremyletang**

### CODE REVIEW

**Review by wwestgarth — DISMISSED** [review]

File: datanode/gateway/middleware.go [hunk 1]
```diff
@@ -124,3 +132,92 @@ func (i *InjectableResponseWriter) Write(data []byte) (int, error) {
...
+func (s *SubscriptionRateLimiter) Dec(ip string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	cnt := s.m[ip]
+	s.m[ip] = cnt - 1
+}
+
+func (s *SubscriptionRateLimiter) WithSubscriptionRateLimiter(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// is that a subscription?
```
Comment (zale144) [inline datanode/gateway/middleware.go:175]: It's just to prevent panic going forward

File: datanode/gateway/middleware.go [hunk 2]
```diff
@@ -124,3 +132,92 @@ func (i *InjectableResponseWriter) Write(data []byte) (int, error) {
...
+		if _, ok := w.(http.Hijacker); !ok {
+			next.ServeHTTP(w, r)
+			return
+		}
+
+		if ip, err := getIP(r); err != nil {
+			s.log.Debug("couldn't get client ip", logging.Error(err))
+		} else {
+			if err := s.Inc(ip); err != nil {
+				s.log.Error("couldn't client reached max sub", logging.Error(err))
```
Comment (zale144) [inline datanode/gateway/middleware.go:185]: This could be a bit clearer

File: datanode/gateway/middleware.go [hunk 3]
```diff
@@ -124,3 +132,92 @@ func (i *InjectableResponseWriter) Write(data []byte) (int, error) {
...
+				w.Write([]byte(err.Error()))
+				// write error
+				return
+			}
+			defer func() {
+				s.Dec(ip)
+			}()
+		}
+
+		next.ServeHTTP(w, r)
```
Comment (zale144) [inline datanode/gateway/middleware.go:196]: Maybe if this was in the defer, the if/else above could be simplified a bit?

**Review by zale144 — COMMENTED** [review]
Just some minor comments, not a show stopper

**Comment by zale144** [comment]
> We could use the rate package instead of re-implementing this. What do you think ?
> 
> 
> 
> https://pkg.go.dev/golang.org/x/time/rate

That one doesn't have any grouping  (e.g. per ip) capability.

**Comment by zale144** [comment]
> > We could use the rate package instead of re-implementing this. What do you think ?
> > 
> > 
> > 
> > https://pkg.go.dev/golang.org/x/time/rate
> 
> That one doesn't have any grouping  (e.g. per ip) capability.

Although, we could have a map of those Limiters, like `ip:limiter`. But it might be a bit heavy on the memory.
Better just take bits from it, like the token-based limiting, and add it to this one.

File: datanode/gateway/middleware.go (hunk 2 above)
Comment (jeremyletang) [reply datanode/gateway/middleware.go:185]: haha I don't what I was trying to say here to be fair.

File: datanode/gateway/middleware.go (hunk 3 above)
Comment (jeremyletang) [reply datanode/gateway/middleware.go:197]: hum not really as it makes weird to check if we've been limited. I'd rather keep like that just now I guess.

File: datanode/gateway/middleware.go (hunk 2 above)
Comment (jeremyletang) [reply datanode/gateway/middleware.go]: ddressed

**Review by zale144 — APPROVED** [review]
Lgtm

**Review by pscott31 — APPROVED** [review]

**Pull request merged**